4032 migrate to ahoy

entantoencuanto opened the pull request:
Closes #4032 


## :v: What does this PR do?

* Removes mixpanel and google analytics tracking code (except if analytics is configured on site)
* Adds and configure ahoy to track each request creating an event by site. This includes adding a column to the events table and an association with Site model.
* Adds sidekiq-bounce and defines a `GobiertoCommon::EventCreatorJob` to create ahoy events which is used from controller concern for each request
* The event stores info of the request and its params filtering some. The ahoy config file also includes some GDPR compliance settings and:
  * Masks IPs on visits
  * Disables use of cookies by ahoy

## :mag: How should this be manually tested?


## :eyes: Screenshots

### Before this PR

### After this PR

## :shipit: Does this PR changes any configuration file?

- [ ] new environment variable in `.env.example`?
- [x] deleted entries in `config/application.yml`?
- [ ] new entry in `config/secrets.yml`?

(Changes in these files might need to update the role in Ansible)

## :book: Does this PR require updating the documentation?

Pending

entantoencuanto committed:
Remove mixpanel references

entantoencuanto committed:
Remove google analytics and use ga only if configured on site settings

entantoencuanto committed:
Add ahoy dependency

entantoencuanto committed:
Update Gemfile.lock

entantoencuanto committed:
Add files included by ahoy installation and association with site

entantoencuanto committed:
Track requests with ahoy

furilo commented on app/controllers/concerns/application_concern.rb:
```diff
@@ -56,6 +61,12 @@ def authenticate_user_in_site
+  def track(name, properties = {})
+    ahoy.track name, **properties.merge(site_id: current_site.id)
+  rescue StandardError => e
+    Rollbar.error(e)
```
@entantoencuanto Rollbar?

entantoencuanto replied:
I've added the rescue to avoid exceptions trying to generate ahoy events, but the use of Rollbar after the rescue is optional

furilo commented:
But we are no longer using Rollbar (or won't soon), but Appsignal
instead...

entantoencuanto committed:
Use Appsignal instead of Rollbar

ferblape commented:
FYI, we still have rollbar because of the JS https://github.com/PopulateTools/gobierto/issues/3240. I'm going to move this issue to Business as usual

entantoencuanto committed:
Add sidekiq-bouncer dependency

entantoencuanto committed:
Configure bouncer and define EventBouncer

entantoencuanto committed:
Add EventCreatorJob and use it in application concern

entantoencuanto committed:
Add event_creators queue to sidekiq configuration

entantoencuanto committed:
Add GDPR compliance settings

entantoencuanto committed:
Remove useless comment

ferblape commented on app/controllers/concerns/application_concern.rb:
```diff
@@ -56,10 +59,29 @@ def authenticate_user_in_site
+  end
+
+  def ignore_tracking_request?
+    Rails.env.development?
```
Add Rails.env.test?

ferblape reviewed (changes requested):
Looks good, just two things:

- should it be tested in some way
- remember to add the ansible PR and update the documentation in regards to the modifications of the `config/application.yml`

entantoencuanto committed:
Ignore ahoy requests tracking on test environment

entantoencuanto committed:
Don't use bouncer in events creator job

entantoencuanto committed:
Make association of events whith visits optional

entantoencuanto committed:
Remove bouncer

entantoencuanto committed:
Add integration tests

ferblape commented:
Hmmm, this is not working well, if you check the latest events from ahoy_events table they are without visit_id nor site_id (I'm generating them visiting Getafe staging site). Please review that

```
-[ RECORD 1 ]---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------
id         | 605
visit_id   |
user_id    |
site_id    |
name       | gobierto_people/people#show
properties | {"slug": "rafael-garcia", "action": "show", "locale": "es", "method": "GET", "controller": "gobierto_people/people"}
time       | 2022-02-17 05:01:23.547261
```

furilo commented:
It's not the same as in contratos Edu?

entantoencuanto committed:
Add current_visit and current_site arguments to EventCreatorJob

entantoencuanto committed:
Update user_id on associated visit after event creation

entantoencuanto commented:
I made a change to not use bounce, and by mistake it was not saving the site correctly for visits with not authenticated users, but the schema of events and method to name them (controller#action) are the same. The only difference is that instead of using an `organization_id` gobierto events use a `site_id`, I can change it.

After talking with Victor I am using a different way of associating visits and events that I think can also be useful in contratos: At the moment of enqueuing the job to create the event from the controller there is a current_visit available which is passed to the job

ferblape commented:
@entantoencuanto thanks, now ahoy_events are tracking the site_id properly.

Would it be possible to add the site_id also to the field ahoy_visits? If we want to give the metric unique visitors to a single site right now it's not possible (please confirm @furilo).

furilo commented:
Yes, that is basic.

entantoencuanto commented:
Yes, it's possible, but  currently the same visit can use different site_ids. For example, currently if I change in staging of site from the same browser the visit remains the same. I can investigate if a different visit can be generated for each site

Or may I do the same as I did with the user_id field, I can update the site_id when it changes

ferblape commented:
If the same user visits two sites it should be tracked as two visits, because sites should be independent. Customers don't care if the application is multitenant or there are N applications.

furilo commented:
Since a tenant only sees his visits, I don't think it's a problem thats
shared. I would not complicate it

entantoencuanto committed:
Add migration to include a reference to site in visits

entantoencuanto committed:
Add association of Ahoy::Visit model with Site

entantoencuanto committed:
Include site in visits tracked by ahoy

This works when Ahoy.cookies is set to false. If cookies are used
the methods to track with cookies should be ovewritten. This hack
work by using some classes inherited from Ahoy ones which includes
site by:

* Defining an Ahoy::GobiertoTracker inheriting from Ahoy::Tracker wich
  includes site in visit_anonimity_set and visitor_anonimity_set methods
  (used to detect if visit exists previously)
* Ovewriting visit and track_visit in Ahoy::Store class to include site

entantoencuanto committed:
Update ahoy tracking integration test

ferblape reviewed (approved):
Now it works as expected!

ferblape commented:
@entantoencuanto there's something I missed in the acceptance that is broken. Your PR doesn't work when a customer has a google analytics key setup (the first point of your PR description). See that you are removing the `ga` function, while it's necessary: https://rollbar.com/Populate/gobierto/items/6057/?utm_campaign=exp_repeat_item&utm_medium=email&utm_source=rollbar-notification&utm_content=control
